File: app/src/main/java/org/nexc/features/meals/RecipesLibraryScreen.kt
/*
 * SPDX-License-Identifier: GPL-3.0-or-later
 * Copyright (c) 2025-2026. The Nexc Contributors
 *
 * Nexc is subject to additional terms covering author attribution and trademark usage;
 * see the ADDITIONAL_TERMS.md and TRADEMARK_POLICY.md files in the project root.
 */

package org.nexc.features.meals

import androidx.compose.foundation.clickable
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.hilt.lifecycle.viewmodel.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import org.nexc.R
import org.nexc.core.components.NexcLazyColumn
import org.nexc.core.components.NexcScaffold
import org.nexc.core.db.entity.Product
import org.nexc.core.db.entity.Recipe
import org.nexc.core.db.entity.RecipeIngredient
import org.nexc.core.db.relations.RecipeIngredientWithProduct
import org.nexc.core.db.relations.RecipeWithIngredients
import java.util.Locale

@Composable
fun RecipesLibraryScreen(
    navigateBack: () -> Unit
) {
    val viewModel: RecipesLibraryViewModel = hiltViewModel()
    val recipes by viewModel.recipes.collectAsStateWithLifecycle()
    val products by viewModel.products.collectAsStateWithLifecycle()

    var showAddEditDialog by remember { mutableStateOf<RecipeWithIngredients?>(null) }

    NexcScaffold(
        title = AnnotatedString(stringResource(R.string.recipes)),
        navigateBack = navigateBack,
        fabAction = { showAddEditDialog = RecipeWithIngredients(Recipe(id = 0L), emptyList()) },
        fabText = stringResource(R.string.add_recipe),
        fabIcon = painterResource(R.drawable.ic_add)
    ) { innerPadding ->
        if (recipes.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(innerPadding),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = stringResource(R.string.no_recipes),
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        } else {
            NexcLazyColumn(innerPadding = innerPadding) {
                items(recipes, key = { it.recipe.id }) { recipe ->
                    RecipeCard(
                        recipeWithIngredients = recipe,
                        onClick = { showAddEditDialog = recipe },
                        onDelete = { viewModel.deleteRecipe(recipe.recipe) }
                    )
                }
            }
        }

        showAddEditDialog?.let { recipeWithIngredients ->
            AddEditRecipeDialog(
                recipeWithIngredients = recipeWithIngredients,
                products = products,
                onDismiss = { showAddEditDialog = null },
                onConfirm = { editedRecipe ->
                    viewModel.saveRecipe(editedRecipe)
                    showAddEditDialog = null
                }
            )
        }
    }
}

@Composable
fun RecipeCard(
    recipeWithIngredients: RecipeWithIngredients,
    onClick: () -> Unit,
    onDelete: () -> Unit
) {
    val recipe = recipeWithIngredients.recipe
    val ingredients = recipeWithIngredients.ingredients

    // Totals calculations
    var totalProt = 0.0
    var totalCarb = 0.0
    var totalFat = 0.0
    var totalCost = 0.0
    ingredients.forEach { item ->
        val scale = item.ingredient.amount / 100.0
        totalProt += item.product.proteins * scale
        totalCarb += item.product.carbs * scale
        totalFat += item.product.fats * scale
        // cost proportion: product.cost is for product.weight
        val costFactor = if (item.product.weight > 0) item.ingredient.amount / item.product.weight else 0.0
        totalCost += item.product.cost * costFactor
    }

    ElevatedCard(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp)
            .clickable { onClick() },
        shape = MaterialTheme.shapes.medium
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = recipe.name,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    SuggestionChip(
                        onClick = {},
                        label = {
                            Text(
                                if (recipe.isPortable) stringResource(R.string.portable)
                                else stringResource(R.string.not_portable)
                            )
                        },
                        colors = SuggestionChipDefaults.suggestionChipColors(
                            containerColor = if (recipe.isPortable) MaterialTheme.colorScheme.primaryContainer else MaterialTheme.colorScheme.errorContainer,
                            labelColor = if (recipe.isPortable) MaterialTheme.colorScheme.onPrimaryContainer else MaterialTheme.colorScheme.onErrorContainer
                        )
                    )
                }
                IconButton(onClick = onDelete) {
                    Icon(
                        painter = painterResource(R.drawable.ic_delete),
                        contentDescription = stringResource(R.string.delete),
                        tint = MaterialTheme.colorScheme.error
                    )
                }
            }

            if (ingredients.isNotEmpty()) {
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = ingredients.joinToString(", ") { "${it.product.name} (${it.ingredient.amount}g)" },
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            if (recipe.instructions.isNotBlank()) {
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = recipe.instructions,
                    style = MaterialTheme.typography.bodyMedium,
                    maxLines = 2
                )
            }

            Spacer(modifier = Modifier.height(8.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "Macros: P ${String.format(Locale.getDefault(), "%.1f", totalProt)}g | C ${String.format(Locale.getDefault(), "%.1f", totalCarb)}g | F ${String.format(Locale.getDefault(), "%.1f", totalFat)}g",
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.SemiBold
                )
                Text(
                    text = "Cost: $${String.format(Locale.getDefault(), "%.2f", totalCost)}",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.secondary,
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddEditRecipeDialog(
    recipeWithIngredients: RecipeWithIngredients,
    products: List<Product>,
    onDismiss: () -> Unit,
    onConfirm: (RecipeWithIngredients) -> Unit
) {
    var name by remember { mutableStateOf(recipeWithIngredients.recipe.name) }
    var instructions by remember { mutableStateOf(recipeWithIngredients.recipe.instructions) }
    var isPortable by remember { mutableStateOf(recipeWithIngredients.recipe.isPortable) }

    val ingredientsList = remember {
        mutableStateListOf<RecipeIngredientWithProduct>().apply {
            addAll(recipeWithIngredients.ingredients)
        }
    }

    var selectedProduct by remember { mutableStateOf<Product?>(null) }
    var amountString by remember { mutableStateOf("") }
    var showProductDropdown by remember { mutableStateOf(false) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                text = if (recipeWithIngredients.recipe.id == 0L) stringResource(R.string.add_recipe) else stringResource(R.string.edit)
            )
        },
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text(stringResource(R.string.name)) },
                    modifier = Modifier.fillMaxWidth()
                )

                OutlinedTextField(
                    value = instructions,
                    onValueChange = { instructions = it },
                    label = { Text(stringResource(R.string.process)) },
                    minLines = 2,
                    modifier = Modifier.fillMaxWidth()
                )

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        checked = isPortable,
                        onCheckedChange = { isPortable = it }
                    )
                    Text(text = stringResource(R.string.is_portable))
                }

                HorizontalDivider()

                Text(
                    text = stringResource(R.string.composition),
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold
                )

                // Add Ingredient sub-form
                Box(modifier = Modifier.fillMaxWidth()) {
                    OutlinedButton(
                        onClick = { showProductDropdown = true },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text(selectedProduct?.name ?: "Select Product/Ingredient")
                    }
                    DropdownMenu(
                        expanded = showProductDropdown,
                        onDismissRequest = { showProductDropdown = false }
                    ) {
                        products.forEach { product ->
                            DropdownMenuItem(
                                text = { Text(product.name) },
                                onClick = {
                                    selectedProduct = product
                                    showProductDropdown = false
                                }
                            )
                        }
                    }
                }

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    OutlinedTextField(
                        value = amountString,
                        onValueChange = { amountString = it },
                        label = { Text("Amount (g/units)") },
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                        modifier = Modifier.weight(1f)
                    )
                    Button(
                        onClick = {
                            val prod = selectedProduct
                            val amt = amountString.toDoubleOrNull()
                            if (prod != null && amt != null && amt > 0.0) {
                                ingredientsList.add(
                                    RecipeIngredientWithProduct(
                                        ingredient = RecipeIngredient(
                                            productId = prod.id,
                                            amount = amt
                                        ),
                                        product = prod
                                    )
                                )
                                amountString = ""
                                selectedProduct = null
                            }
                        },
                        enabled = selectedProduct != null && amountString.isNotBlank()
                    ) {
                        Text(stringResource(R.string.add))
                    }
                }

                // List added ingredients
                ingredientsList.forEach { ing ->
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = "${ing.product.name}: ${ing.ingredient.amount}g",
                            style = MaterialTheme.typography.bodyMedium
                        )
                        IconButton(onClick = { ingredientsList.remove(ing) }) {
                            Icon(
                                painter = painterResource(R.drawable.ic_delete),
                                contentDescription = stringResource(R.string.delete),
                                tint = MaterialTheme.colorScheme.error
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            Button(
                onClick = {
                    if (name.isNotBlank()) {
                        val finalRecipe = Recipe(
                            id = if (recipeWithIngredients.recipe.id == 0L) kotlin.random.Random.nextLong() else recipeWithIngredients.recipe.id,
                            name = name,
                            instructions = instructions,
                            isPortable = isPortable
                        )
                        onConfirm(
                            RecipeWithIngredients(
                                recipe = finalRecipe,
                                ingredients = ingredientsList.toList()
                            )
                        )
                    }
                }
            ) {
                Text(stringResource(R.string.save))
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text(stringResource(R.string.cancel_dialog))
            }
        }
    )
}

File: lib/features/meals/recipes_library_screen.dart
/*
 * SPDX-License-Identifier: GPL-3.0-or-later
 * Copyright (c) 2026. The LibreFit Contributors
 * Copyright (c) 2026. The Nexc Contributors
 *
 * LibreFit is subject to additional terms covering author attribution and trademark usage;
 * see the ADDITIONAL_TERMS.md and TRADEMARK_POLICY.md files in the project root.
 */

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:collection/collection.dart';
import '../../core/db/relations.dart';
import '../../core/db/meal_repository.dart';
import '../../core/providers/meals_providers.dart';
import '../../core/components/nexc_scaffold.dart';

class RecipesLibraryScreen extends ConsumerStatefulWidget {
  const RecipesLibraryScreen({super.key});

  @override
  ConsumerState<RecipesLibraryScreen> createState() => _RecipesLibraryScreenState();
}

class _RecipesLibraryScreenState extends ConsumerState<RecipesLibraryScreen> {
  String _searchQuery = "";

  void _showAddEditRecipe(BuildContext context, RecipeWithIngredients recipe, List<Product> products) {
    showDialog(
      context: context,
      barrierDismissible: false,
      builder: (context) => AddEditRecipeDialog(
        recipeWithIngredients: recipe,
        products: products,
        onDismiss: () => Navigator.pop(context),
        onConfirm: (editedRecipe) {
          ref.read(mealRepositoryProvider).saveRecipeWithIngredients(editedRecipe);
          Navigator.pop(context);
        },
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final recipesAsync = ref.watch(allRecipesProvider);
    final products = ref.watch(allProductsProvider).value ?? [];
    final theme = Theme.of(context);

    return NexcScaffold(
      title: const Text("Recipes"),
      navigateBack: () => Navigator.pop(context),
      fabAction: () {
        _showAddEditRecipe(
          context,
          RecipeWithIngredients(
            recipe: Recipe(
              id: 0,
              name: _searchQuery.trim(),
              instructions: "",
              isPortable: true,
            ),
            ingredients: [],
          ),
          products,
        );
      },
      fabIcon: const Icon(Icons.add),
      fabText: "Add Recipe",
      content: (context, padding) {
        return recipesAsync.when(
          loading: () => const Center(child: CircularProgressIndicator()),
          error: (err, stack) => Center(child: Text("Error: $err")),
          data: (recipes) {
            final filteredRecipes = _searchQuery.isEmpty
                ? recipes
                : recipes.where((r) {
                    final nameMatch = r.recipe.name.toLowerCase().contains(_searchQuery.toLowerCase());
                    final ingredientMatch = r.ingredients.any(
                        (ing) => ing.product.name.toLowerCase().contains(_searchQuery.toLowerCase()));
                    return nameMatch || ingredientMatch;
                  }).toList();

            return Column(
              children: [
                Padding(
                  padding: const EdgeInsets.all(16.0),
                  child: TextField(
                    decoration: const InputDecoration(
                      labelText: "Search recipes...",
                      prefixIcon: Icon(Icons.search),
                      border: OutlineInputBorder(
                        borderRadius: BorderRadius.all(Radius.circular(12.0)),
                      ),
                    ),
                    onChanged: (val) {
                      setState(() {
                        _searchQuery = val;
                      });
                    },
                  ),
                ),
                Expanded(
                  child: filteredRecipes.isEmpty
                      ? const Center(child: Text("No recipes found"))
                      : ListView.builder(
                          padding: const EdgeInsets.only(left: 16.0, right: 16.0, bottom: 80.0),
                          itemCount: filteredRecipes.length,
                          itemBuilder: (context, index) {
                            final recipe = filteredRecipes[index];
                            return Padding(
                              key: ValueKey(recipe.recipe.id),
                              padding: const EdgeInsets.only(bottom: 8.0),
                              child: RecipeCard(
                                recipeWithIngredients: recipe,
                                onClick: () {
                                  _showAddEditRecipe(context, recipe, products);
                                },
                                onDelete: () async {
                                  final confirm = await showDialog<bool>(
                                    context: context,
                                    builder: (context) => AlertDialog(
                                      title: const Text("Delete Recipe"),
                                      content: Text("Are you sure you want to delete '${recipe.recipe.name}'?"),
                                      actions: [
                                        TextButton(
                                          onPressed: () => Navigator.pop(context, false),
                                          child: const Text("Cancel"),
                                        ),
                                        TextButton(
                                          onPressed: () => Navigator.pop(context, true),
                                          child: Text("Delete", style: TextStyle(color: theme.colorScheme.error)),
                                        ),
                                      ],
                                    ),
                                  );
                                  if (confirm == true) {
                                    ref.read(mealRepositoryProvider).deleteRecipe(recipe.recipe);
                                  }
                                },
                              ),
                            );
                          },
                        ),
                ),
              ],
            );
          },
        );
      },
    );
  }
}

class RecipeCard extends StatelessWidget {
  final RecipeWithIngredients recipeWithIngredients;
  final VoidCallback onClick;
  final VoidCallback onDelete;

  const RecipeCard({
    super.key,
    required this.recipeWithIngredients,
    required this.onClick,
    required this.onDelete,
  });

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final recipe = recipeWithIngredients.recipe;
    final ingredients = recipeWithIngredients.ingredients;

    // Totals calculations
    double totalProt = 0.0;
    double totalCarb = 0.0;
    double totalFat = 0.0;
    for (final item in ingredients) {
      final scale = item.ingredient.amount / 100.0;
      totalProt += item.product.proteins * scale;
      totalCarb += item.product.carbs * scale;
      totalFat += item.product.fats * scale;
    }

    return Card(
      child: InkWell(
        onTap: onClick,
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          recipe.name,
                          style: theme.textTheme.titleMedium?.copyWith(
                                fontWeight: FontWeight.bold,
                              ),
                        ),
                        const SizedBox(height: 4),
                        Container(
                          padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                          decoration: BoxDecoration(
                            color: recipe.isPortable
                                ? theme.colorScheme.primaryContainer.withOpacity(0.5)
                                : theme.colorScheme.errorContainer.withOpacity(0.5),
                            borderRadius: BorderRadius.circular(4),
                          ),
                          child: Text(
                            recipe.isPortable ? "Portable" : "Home only",
                            style: TextStyle(
                              fontSize: 10,
                              color: recipe.isPortable
                                  ? theme.colorScheme.onPrimaryContainer
                                  : theme.colorScheme.onErrorContainer,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ),
                      ],
                    ),
                  ),
                  IconButton(
                    icon: Icon(Icons.delete_outline, color: theme.colorScheme.error),
                    onPressed: onDelete,
                  ),
                ],
              ),
              if (ingredients.isNotEmpty) ...[
                const SizedBox(height: 4),
                Text(
                  ingredients.map((i) => "${i.product.name} (${i.ingredient.amount.toStringAsFixed(0)}${i.ingredient.amountUnits ?? 'g'})").join(", "),
                  style: theme.textTheme.bodySmall?.copyWith(
                        color: theme.colorScheme.onSurfaceVariant,
                      ),
                ),
              ],
              if (recipe.instructions.isNotEmpty) ...[
                const SizedBox(height: 8),
                Text(
                  recipe.instructions,
                  style: theme.textTheme.bodyMedium,
                  maxLines: 2,
                  overflow: TextOverflow.ellipsis,
                ),
              ],
              const SizedBox(height: 8),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Text(
                    "Macros: P ${totalProt.toStringAsFixed(1)}g | C ${totalCarb.toStringAsFixed(1)}g | F ${totalFat.toStringAsFixed(1)}g",
                    style: theme.textTheme.bodySmall?.copyWith(fontWeight: FontWeight.bold),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class AddEditRecipeDialog extends StatefulWidget {
  final RecipeWithIngredients recipeWithIngredients;
  final List<Product> products;
  final VoidCallback onDismiss;
  final void Function(RecipeWithIngredients) onConfirm;

  const AddEditRecipeDialog({
    super.key,
    required this.recipeWithIngredients,
    required this.products,
    required this.onDismiss,
    required this.onConfirm,
  });

  @override
  State<AddEditRecipeDialog> createState() => _AddEditRecipeDialogState();
}

class _AddEditRecipeDialogState extends State<AddEditRecipeDialog> {
  late TextEditingController _nameController;
  late TextEditingController _instructionsController;
  late bool _isPortable;
  late List<RecipeIngredientWithProduct> _ingredients;

  int? _selectedProductId;
  final TextEditingController _amountController = TextEditingController();
  String _selectedIngredientUnit = "g";
  String _searchQuery = "";

  @override
  void initState() {
    super.initState();
    _selectedIngredientUnit = "g";
    _nameController = TextEditingController(text: widget.recipeWithIngredients.recipe.name);
    _instructionsController =
        TextEditingController(text: widget.recipeWithIngredients.recipe.instructions);
    _isPortable = widget.recipeWithIngredients.recipe.isPortable;
    _ingredients = List.from(widget.recipeWithIngredients.ingredients);
  }

  @override
  void dispose() {
    _nameController.dispose();
    _instructionsController.dispose();
    _amountController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final filteredProducts = _searchQuery.isEmpty
        ? widget.products
        : widget.products.searchAndSort(_searchQuery);

    return AlertDialog(
      title: Text(widget.recipeWithIngredients.recipe.id == 0 ? "Add Recipe" : "Edit Recipe"),
      content: SizedBox(
        width: 320,
        child: SingleChildScrollView(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              TextField(
                controller: _nameController,
                decoration: const InputDecoration(labelText: "Name"),
              ),
              const SizedBox(height: 8),
              TextField(
                controller: _instructionsController,
                decoration: const InputDecoration(labelText: "Instructions/Process"),
                minLines: 2,
                maxLines: 5,
              ),
              const SizedBox(height: 8),
              Row(
                children: [
                  Checkbox(
                    value: _ingredients.any((ing) => !ing.product.isPortable) ? false : _isPortable,
                    onChanged: _ingredients.any((ing) => !ing.product.isPortable)
                        ? null
                        : (val) {
                            if (val != null) setState(() => _isPortable = val);
                          },
                  ),
                  Expanded(
                    child: Text(
                      _ingredients.any((ing) => !ing.product.isPortable)
                          ? "Is Portable (Contains non-portable ingredients)"
                          : "Is Portable",
                      style: TextStyle(
                        color: _ingredients.any((ing) => !ing.product.isPortable) ? theme.disabledColor : null,
                      ),
                    ),
                  ),
                ],
              ),
              const Divider(),
              const Text("Composition", style: TextStyle(fontWeight: FontWeight.bold)),
              const SizedBox(height: 8),
              if (_selectedProductId != null) ...[
                Builder(builder: (context) {
                  final prod = widget.products.firstWhereOrNull((p) => p.id == _selectedProductId);
                  return Card(
                    margin: const EdgeInsets.symmetric(vertical: 8),
                    child: ListTile(
                      leading: const Icon(Icons.check_circle, color: Colors.green),
                      title: Text(prod?.name ?? ""),
                      subtitle: Text("${prod?.proteins}g P | ${prod?.carbs}g C | ${prod?.fats}g F"),
                      trailing: IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: () => setState(() => _selectedProductId = null),
                      ),
                    ),
                  );
                }),
              ] else ...[
                TextField(
                  decoration: const InputDecoration(
                    labelText: "Search Ingredient...",
                    prefixIcon: Icon(Icons.search),
                    border: OutlineInputBorder(),
                  ),
                  onChanged: (val) => setState(() => _searchQuery = val),
                ),
                const SizedBox(height: 8),
                ConstrainedBox(
                  constraints: const BoxConstraints(maxHeight: 180),
                  child: Card(
                    elevation: 0,
                    shape: RoundedRectangleBorder(
                      side: BorderSide(color: theme.colorScheme.outlineVariant),
                      borderRadius: BorderRadius.circular(8),
                    ),
                    child: ListView.builder(
                      shrinkWrap: true,
                      itemCount: filteredProducts.length,
                      itemBuilder: (context, idx) {
                        final p = filteredProducts[idx];
                        return ListTile(
                          title: Text(p.name),
                          subtitle: Text("${p.proteins}g P | ${p.carbs}g C | ${p.fats}g F"),
                          onTap: () => setState(() {
                            _selectedProductId = p.id;
                            _searchQuery = "";
                          }),
                        );
                      },
                    ),
                  ),
                ),
              ],
              const SizedBox(height: 8),
              Row(
                children: [
                  Expanded(
                    child: TextField(
                      controller: _amountController,
                      keyboardType: const TextInputType.numberWithOptions(decimal: true),
                      decoration: const InputDecoration(labelText: "Amount"),
                    ),
                  ),
                   const SizedBox(width: 8),
                  Builder(
                    builder: (context) {
                      final prod = widget.products.firstWhereOrNull((p) => p.id == _selectedProductId);
                      final showMl = prod == null || (prod.mlToGFactor != null && prod.mlToGFactor! > 0);
                      final showUnits = prod == null || (prod.unitWeight != null && prod.unitWeight! > 0);
                      if (_selectedIngredientUnit == "ml" && !showMl) {
                        _selectedIngredientUnit = "g";
                      }
                      if (_selectedIngredientUnit == "units" && !showUnits) {
                        _selectedIngredientUnit = "g";
                      }
                      return DropdownButton<String>(
                        value: _selectedIngredientUnit,
                        items: [
                          const DropdownMenuItem(value: "g", child: Text("g")),
                          if (showMl)
                            const DropdownMenuItem(value: "ml", child: Text("ml")),
                          if (showUnits)
                            const DropdownMenuItem(value: "units", child: Text("units")),
                        ],
                        onChanged: (val) {
                          if (val != null) {
                            setState(() {
                              _selectedIngredientUnit = val;
                            });
                          }
                        },
                      );
                    },
                  ),
                  const SizedBox(width: 8),
                  ElevatedButton(
                    onPressed: () {
                      final prod = widget.products.firstWhereOrNull((p) => p.id == _selectedProductId);
                      final amt = double.tryParse(_amountController.text) ?? 0.0;
                      if (prod != null && amt > 0.0) {
                        setState(() {
                          _ingredients.add(
                            RecipeIngredientWithProduct(
                              ingredient: RecipeIngredient(
                                id: 0,
                                recipeId: widget.recipeWithIngredients.recipe.id,
                                productId: prod.id,
                                amount: amt,
                                amountUnits: _selectedIngredientUnit,
                              ),
                              product: prod,
                            ),
                          );
                          _amountController.clear();
                          _selectedProductId = null;
                        });
                      }
                    },
                    child: const Text("Add"),
                  ),
                ],
              ),
              const SizedBox(height: 12),
              ..._ingredients.map((ing) {
                return Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Text("${ing.product.name}: ${ing.ingredient.amount.toStringAsFixed(0)}${ing.ingredient.amountUnits ?? 'g'}"),
                    IconButton(
                      icon: Icon(Icons.delete_outline, color: theme.colorScheme.error),
                      onPressed: () {
                        setState(() {
                          _ingredients.remove(ing);
                        });
                      },
                    ),
                  ],
                );
              }),
            ],
          ),
        ),
      ),
      actions: [
        IconButton(
          icon: const Icon(Icons.close),
          onPressed: widget.onDismiss,
        ),
        IconButton(
          icon: const Icon(Icons.check),
          onPressed: () {
            final name = _nameController.text.trim();
            if (name.isNotEmpty) {
              final newRecipe = Recipe(
                id: widget.recipeWithIngredients.recipe.id,
                name: name,
                instructions: _instructionsController.text.trim(),
                isPortable: _isPortable,
              );
              widget.onConfirm(
                RecipeWithIngredients(recipe: newRecipe, ingredients: _ingredients),
              );
            }
          },
        ),
      ],
    );
  }
}
